node-ipc/src/Cardano/NodeIPC.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wall            #-}
+
+module Cardano.NodeIPC (startNodeJsIPC) where
+
+import           Control.Arrow             ( (>>>) )
+import           Control.Concurrent
+import           Control.Exception ()
+import           Control.Monad.Reader (MonadReader)
+import           Control.Monad.IO.Class
+import           Data.Aeson
+import           Data.Aeson.Types          (Options, SumEncoding(ObjectWithSingleField), sumEncoding)
+import           Data.Binary.Get           (runGet, getWord64le, getWord32le)
+import           Data.Binary.Put           (runPut, putWord32le, putWord64le, putLazyByteString)
+import qualified Data.ByteString.Lazy      as BSL
+import qualified Data.ByteString.Lazy.Char8 as BSLC
+import           Distribution.System       (buildOS, OS(Windows))
+import           Foreign.C.Types           (CInt)
+import           GHC.IO.Handle.FD          (fdToHandle)
+import           GHC.Generics              (Generic)
+import           Pos.Shutdown.Logic        (triggerShutdown)
+import           Pos.Shutdown.Class        (HasShutdownContext (..))
+import           Pos.Shutdown.Types        (ShutdownContext)
+import           System.Environment        (lookupEnv)
+import           System.IO.Error           (IOError, isEOFError)
+import           System.IO                 (hFlush, hGetLine, hSetNewlineMode, noNewlineTranslation, hGetChar)
+import           System.Wlog               (WithLogger, logInfo, logError)
+import           System.Wlog.LoggerNameBox (usingLoggerName)
+import           Universum
+
+data Packet = Started | QueryPort | ReplyPort Word16 | Ping | Pong | ParseError Text deriving (Show, Eq, Generic)
+
+opts :: Options
+opts = defaultOptions { sumEncoding = ObjectWithSingleField }
+
+instance FromJSON Packet where
+  parseJSON = genericParseJSON opts
+
+instance ToJSON Packet where
+  toEncoding = genericToEncoding opts
+
+startNodeJsIPC ::
+    (MonadIO m, WithLogger m, MonadReader ctx m, HasShutdownContext ctx)
+    => Word16 -> m ()
+startNodeJsIPC port = void $ runMaybeT $ do
+  ctx <- view shutdownContext
+  fdstring <- liftIO (lookupEnv "NODE_CHANNEL_FD") >>= (pure >>> MaybeT)
+  case readEither fdstring of
+    Left err -> lift $ logError $ "unable to parse NODE_CHANNEL_FD: " <> err
+    Right fd -> void $ liftIO $ forkIO $ startIpcListener ctx fd port
+
+startIpcListener :: ShutdownContext -> CInt -> Word16 -> IO ()
+startIpcListener ctx fd port = usingLoggerName "NodeIPC" $ flip runReaderT ctx (ipcListener fd port)
+
+ipcListener ::
+    forall m ctx . (MonadCatch m, MonadIO m, WithLogger m, MonadReader ctx m, HasShutdownContext ctx)
+    => CInt -> Word16 -> m ()
+ipcListener fd port = do
```

Comment:
Couple of things here.

Firstly this doesn't conform to the coding standards, which specify 4 space indentation rather than 2. I actually prefer 2 space indentation in my own code, but in this project i stick to the convention.

Secondy, this is one huge function with a whole bunch of functions defined in `let` expressions. I think the intention of this code would be a lot clearer if all the small functions were split out, even if that meant that some more explicit parameter passing was needed. Since none of these functions are exported from the module, they are likely to be in-lined by the compiler anyway, but having them out-of-line in the source significantly improves readability.